# Credit Risk Resampling Techniques

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import RandomOverSampler
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import ClusterCentroids
from imblearn.combine import SMOTEENN
from sklearn.metrics import balanced_accuracy_score
from imblearn.metrics import classification_report_imbalanced

# Read the CSV into DataFrame

In [3]:
# Load the data
file_path = Path('Resources/lending_data.csv')
df = pd.read_csv(file_path)
df.head()

loan_size  interest_rate homeowner  borrower_income  debt_to_income  \
0    10700.0          7.672       own            52800        0.431818   
1     8400.0          6.692       own            43600        0.311927   
2     9000.0          6.963      rent            46100        0.349241   
3    10700.0          7.664       own            52700        0.430740   
4    10800.0          7.698  mortgage            53000        0.433962   

   num_of_accounts  derogatory_marks  total_debt loan_status  
0                5                 1       22800    low_risk  
1                3                 0       13600    low_risk  
2                3                 0       16100    low_risk  
3                5                 1       22700    low_risk  
4                5                 1       23000    low_risk

In [4]:
# Create the LabelEncoder instance
le = LabelEncoder()

# Fitting and encoding the columns with the LabelEncoder

# Home Owner column
le.fit(df["homeowner"])
df["homeowner"] = le.transform(df["homeowner"])

df.head()

loan_size  interest_rate  homeowner  borrower_income  debt_to_income  \
0    10700.0          7.672          1            52800        0.431818   
1     8400.0          6.692          1            43600        0.311927   
2     9000.0          6.963          2            46100        0.349241   
3    10700.0          7.664          1            52700        0.430740   
4    10800.0          7.698          0            53000        0.433962   

   num_of_accounts  derogatory_marks  total_debt loan_status  
0                5                 1       22800    low_risk  
1                3                 0       13600    low_risk  
2                3                 0       16100    low_risk  
3                5                 1       22700    low_risk  
4                5                 1       23000    low_risk

# Split the Data into Training and Testing

In [5]:
# Create our features
X = df.drop(columns="loan_status")

# Create our target
y = df["loan_status"]

In [6]:
X.describe()

loan_size  interest_rate     homeowner  borrower_income  \
count  77536.000000   77536.000000  77536.000000     77536.000000   
mean    9805.562577       7.292333      0.606144     49221.949804   
std     2093.223153       0.889495      0.667811      8371.635077   
min     5000.000000       5.250000      0.000000     30000.000000   
25%     8700.000000       6.825000      0.000000     44800.000000   
50%     9500.000000       7.172000      1.000000     48100.000000   
75%    10400.000000       7.528000      1.000000     51400.000000   
max    23800.000000      13.235000      2.000000    105200.000000   

       debt_to_income  num_of_accounts  derogatory_marks    total_debt  
count    77536.000000     77536.000000      77536.000000  77536.000000  
mean         0.377318         3.826610          0.392308  19221.949804  
std          0.081519         1.904426          0.582086   8371.635077  
min          0.000000         0.000000          0.000000      0.000000  
25%          0.330357         3.000000          0.000000  14800.000000  
50%          0.376299         4.000000          0.000000  18100.000000  
75%          0.416342         4.000000          1.000000  21400.000000  
max          0.714829        16.000000          3.000000  75200.000000

In [7]:
y.describe()
y.head()

0    low_risk
1    low_risk
2    low_risk
3    low_risk
4    low_risk
Name: loan_status, dtype: object

In [8]:
# Check the balance of our target values
y.value_counts()

low_risk     75036
high_risk     2500
Name: loan_status, dtype: int64

In [9]:
# Create X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)

X_train.head()

loan_size  interest_rate  homeowner  borrower_income  debt_to_income  \
25729    10100.0          7.402          0            50300        0.403579   
20176     9200.0          7.049          0            46900        0.360341   
55080     8800.0          6.870          0            45200        0.336283   
10410     8900.0          6.919          0            45700        0.343545   
63110     8500.0          6.731          1            43900        0.316629   

       num_of_accounts  derogatory_marks  total_debt  
25729                4                 1       20300  
20176                3                 0       16900  
55080                3                 0       15200  
10410                3                 0       15700  
63110                3                 0       13900

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [10]:
# Create the StandardScaler instance
scaler = StandardScaler()

In [11]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [12]:
# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Simple Logistic Regression

In [13]:
model_slr = LogisticRegression(solver='lbfgs', random_state=1)
model_slr.fit(X_train, y_train)

LogisticRegression(random_state=1)

In [14]:
# Calculated the balanced accuracy score
y_pred = model_slr.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9543211898288821

In [15]:
# Display the confusion matrix
cm_slr = confusion_matrix(y_test, y_pred)
cm_slr_df = pd.DataFrame(
    cm_slr, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
cm_slr_df

Predicted 0  Predicted 1
Actual 0          571           54
Actual 1           93        18666

In [16]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.86      0.91      1.00      0.89      0.95      0.90       625
   low_risk       1.00      1.00      0.91      1.00      0.95      0.92     18759

avg / total       0.99      0.99      0.92      0.99      0.95      0.92     19384



# Oversampling

In this section, you will compare two oversampling algorithms to determine which algorithm results in the best performance. You will oversample the data using the naive random oversampling algorithm and the SMOTE algorithm. For each algorithm, be sure to complete the folliowing steps:

1. View the count of the target classes using `Counter` from the collections library. 
3. Use the resampled data to train a logistic regression model.
3. Calculate the balanced accuracy score from sklearn.metrics.
4. Print the confusion matrix from sklearn.metrics.
5. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.

Note: Use a random state of 1 for each sampling algorithm to ensure consistency between tests

### Naive Random Oversampling

In [17]:
# Resample the training data with the RandomOversampler
ros = RandomOverSampler(random_state=1)
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)

# View the count of target classes with Counter
Counter(y_resampled)

Counter({'low_risk': 56277, 'high_risk': 56277})

In [18]:
# Train the Logistic Regression model using the resampled data
model_nrolr = LogisticRegression(solver='lbfgs', random_state=1)
model_nrolr.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [19]:
# Calculated the balanced accuracy score

balanced_accuracy_score(y_test, y_pred)

0.9543211898288821

In [20]:
# Display the confusion matrix
y_pred = model_nrolr.predict(X_test)
cm_nro = confusion_matrix(y_test, y_pred)
cm_nro_df = pd.DataFrame(
    cm_nro, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
cm_nro_df

Predicted 0  Predicted 1
Actual 0          622            3
Actual 1          105        18654

In [21]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.86      1.00      0.99      0.92      0.99      0.99       625
   low_risk       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       1.00      0.99      1.00      0.99      0.99      0.99     19384



### SMOTE Oversampling

In [22]:
# Resample the training data with SMOTE
X_resampled, y_resampled = SMOTE(random_state=1, sampling_strategy=1.0).fit_resample(
    X_train, y_train
)

# View the count of target classes with Counter
Counter(y_resampled)

Counter({'low_risk': 56277, 'high_risk': 56277})

In [23]:
# Train the Logistic Regression model using the resampled data
model_smote = LogisticRegression(solver='lbfgs', random_state=1)
model_smote.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [24]:
# Calculated the balanced accuracy score
y_pred = model_smote.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9948279972279972

In [25]:
# Display the confusion matrix
cm_smote = confusion_matrix(y_test, y_pred)
cm_smote_df = pd.DataFrame(
    cm_smote, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
cm_smote_df

Predicted 0  Predicted 1
Actual 0          622            3
Actual 1          104        18655

In [26]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.86      1.00      0.99      0.92      0.99      0.99       625
   low_risk       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       1.00      0.99      1.00      0.99      0.99      0.99     19384



# Undersampling

In this section, you will test an undersampling algorithm to determine which algorithm results in the best performance compared to the oversampling algorithms above. You will undersample the data using the Cluster Centroids algorithm and complete the folliowing steps:

1. View the count of the target classes using `Counter` from the collections library. 
3. Use the resampled data to train a logistic regression model.
3. Calculate the balanced accuracy score from sklearn.metrics.
4. Display the confusion matrix from sklearn.metrics.
5. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.

Note: Use a random state of 1 for each sampling algorithm to ensure consistency between tests

In [27]:
# Resample the data using the ClusterCentroids resampler

cc = ClusterCentroids(random_state=1)
X_resampled, y_resampled = cc.fit_resample(X_train, y_train)

# View the count of target classes with Counter
Counter(y_resampled)

Counter({'high_risk': 1875, 'low_risk': 1875})

In [28]:
# Train the Logistic Regression model using the resampled data
model_uslr = LogisticRegression(solver='lbfgs', random_state=1)
model_uslr.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [29]:
# Calculate the balanced accuracy score
balanced_accuracy_score(y_test, y_pred)

0.9948279972279972

In [30]:
# Display the confusion matrix
y_pred = model_uslr.predict(X_test)
cm_cc = confusion_matrix(y_test, y_pred)
cm_cc_df = pd.DataFrame(
    cm_cc, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
cm_cc_df


Predicted 0  Predicted 1
Actual 0          608           17
Actual 1          102        18657

In [31]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.86      0.97      0.99      0.91      0.98      0.97       625
   low_risk       1.00      0.99      0.97      1.00      0.98      0.97     18759

avg / total       0.99      0.99      0.97      0.99      0.98      0.97     19384



# Combination (Over and Under) Sampling

In this section, you will test a combination over- and under-sampling algorithm to determine if the algorithm results in the best performance compared to the other sampling algorithms above. You will resample the data using the SMOTEENN algorithm and complete the folliowing steps:

1. View the count of the target classes using `Counter` from the collections library. 
3. Use the resampled data to train a logistic regression model.
3. Calculate the balanced accuracy score from sklearn.metrics.
4. Display the confusion matrix from sklearn.metrics.
5. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.

Note: Use a random state of 1 for each sampling algorithm to ensure consistency between tests

In [32]:
# Resample the training data with SMOTEENN
smote_enn = SMOTEENN(random_state=0)
X_resampled, y_resampled = smote_enn.fit_resample(X, y)

# View the count of target classes with Counter
Counter(y_resampled)

Counter({'high_risk': 74080, 'low_risk': 74591})

In [33]:
# Train the Logistic Regression model using the resampled data
model_smooteen = LogisticRegression(solver='lbfgs', random_state=1)
model_smooteen.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [34]:
# Calculate the balanced accuracy score
balanced_accuracy_score(y_test, y_pred)

0.9836813049736126

In [35]:
# Display the confusion matrix
y_pred = model_smooteen.predict(X_test)
cm_smoteen = confusion_matrix(y_test, y_pred)
cm_smoteen_df = pd.DataFrame(
    cm_smoteen, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)
cm_smoteen_df

Predicted 0  Predicted 1
Actual 0          622            3
Actual 1          107        18652

In [36]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.85      1.00      0.99      0.92      0.99      0.99       625
   low_risk       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       1.00      0.99      1.00      0.99      0.99      0.99     19384



# Final Questions

1. Which model had the best balanced accuracy score?

   **SMOTE Oversampling and Cluster Centroids under sampling have the best balance accuracy score of 99.48%.**

2. Which model had the best recall score?

   **SMOTE, SMOOTEEN and Naive random oversampling got the best recall score of 1 and .99 for high risk and low risk respectively. Others were pretty close as well.**

3. Which model had the best geometric mean score?

   **SMOTE, SMOOTEEN and Naive random oversampling got the best geometric mean score of 99%.**
